spec/requests/api/v1/assets_controller_spec.rb

```diff
@@ -0,0 +1,157 @@
+        get(api_v1_team_project_experiment_task_protocol_step_assets_path(
+              team_id: @team.id,
+              project_id: @project.id,
+              experiment_id: @experiment.id,
+              task_id: @task.id,
+              protocol_id: -1,
+              step_id: @step.id
+            ), headers: @valid_headers)
+
+        expect(response).to have_http_status(404)
+      end
+    end
+  end
+
+  describe 'GET step, #show' do
+    context 'when has valid params' do
+      it 'renders 200' do
+        create(:step_asset, step: @step, asset: asset)
+
+        get(api_v1_team_project_experiment_task_protocol_step_asset_path(
+              team_id: @team.id,
+              project_id: @project.id,
+              experiment_id: @experiment.id,
+              task_id: @task.id,
+              protocol_id: @protocol.id,
+              step_id: @step.id,
+              id: asset.id
+            ), headers: @valid_headers)
+
+        expect(response).to have_http_status(200)
+      end
+    end
+
+    context 'when experiment is not found' do
+      it 'renders 404' do
+        get(api_v1_team_project_experiment_task_protocol_step_asset_path(
+              team_id: @team.id,
+              project_id: @project.id,
+              experiment_id: -1,
+              task_id: @task.id,
+              protocol_id: @protocol.id,
+              step_id: @step.id,
+              id: asset.id
+            ), headers: @valid_headers)
+
+        expect(response).to have_http_status(404)
+      end
+    end
+  end
+
+  describe 'POST step, #create' do
+    before :all do
+      @valid_headers['Content-Type'] = 'application/json'
+    end
+
+    before :each do
+      @file = fixture_file_upload('files/test.jpg', 'image/jpg')
+    end
+
+    let(:action) do
+      post(api_v1_team_project_experiment_task_protocol_step_assets_path(
+             team_id: @team.id,
+             project_id: @project.id,
+             experiment_id: @experiment.id,
+             task_id: @task.id,
+             protocol_id: @protocol.id,
+             step_id: @step.id
+           ),
+           params: request_body.to_json,
+           headers: @valid_headers)
+    end
+
+    context 'when has valid params' do
+      let(:request_body) do
+        {
+          data: {
+            type: 'attachments',
+            attributes: {
+              file: @file
+            }
+          }
+        }
+      end
+
+      it 'creates new asset' do
+        expect { action }.to change { Asset.count }.by(1)
+      end
+
+      it 'returns status 201' do
+        action
+
+        expect(response).to have_http_status 201
+      end
+    end
+
+    context 'when has missing param' do
+      let(:request_body) do
+        {
+          data: {
+            type: 'attachments',
+            attributes: {
+            }
+          }
+        }
+      end
+
+      it 'renders 400' do
+        action
+
+        expect(response).to have_http_status(400)
+      end
+    end
+  end
```

Comment:
A test that checks asset permission with a wrong user (with no permissions) would be welcomed, since this is quite important.

Reply:
We are checking permission with `can_read_protocol_in_module?(@asset.step.protocol)`
So we check same permission two times with `load_protocol` and `load_step` in the base controller. With the current structure, we won't hit PermissionError on load_asset. 
But asset could be not found, maybe this test makes more sense. 
What do you think? @jbargu

Reply:
Yeah as long as we can confirm that one user cannot access some other asset she cannot have access to I believe we are ok. In case someone changes something that breaches this assumption.